### ISSUE TITLE
volume viewer GUI

### ISSUE THREAD

**Issue by hainm** [arose/nglview#561]
- Support color by volume.

**Comment by jbarnoud** [comment]
Until such gui is implemented, is there a way to set the value to use for the isovolume?

**Comment by hainm** [comment]
hi, in which context? (`update_surface`?)

**Comment by hainm** [comment]
@arose

**Comment by jbarnoud** [comment]
I calculated a 3D density grid. I can display that grid as an isosurface around my molecule of interest, but I would like to choose what density value to use to draw the isosurface.

**Comment by arose** [comment]
Are you loading it with `view.add_component('my.ccp4')`?

@hainm after loading such a component, how can you add representations to it?

**Comment by hainm** [comment]
```python
view.add_surface(..., component=0) # or 1, 2,

# or
view.component_0.add_surface(...)
```

**Comment by arose** [comment]
so `view.component_0.add_surface(threshold=1.0)` should work, I think

**Comment by hainm** [comment]
also update current surface repr.
```python
view.component_0.update_surface(threshold=1.0)
```

**Comment by jbarnoud** [comment]
It works, indeed. Thanks.

Is there a mailing list somewhere to ask questions without polluting the bug tracker?

**Comment by hainm** [comment]
> Is there a mailing list somewhere to ask questions without polluting the bug tracker?

we don't mind about this. I myself prefer to use github issue here. 
Just create a topic for posting random + usage questions: https://github.com/arose/nglview/issues/589

**Comment by arose** [comment]
The amount of questions is not that high, so just having them as separate issues marked as question if fine with me

**Comment by hainm** [comment]
go back to the original issue: demo from @arose 

https://codepen.io/arose/full/oWOQMg/

**Comment by satary** [comment]
Hi @hainm, Hi @arose !
I wonder if it is possible to color a structure (cartoon, surface and other atomic representations) by volume data (loaded with ccp4 or dx).
I see that there is a [volumeColormaker](http://nglviewer.org/ngl/api/class/src/color/volume-colormaker.js~VolumeColormaker.html) but I don't get how to use it. I guess that volume is loaded as a separate component, so can it be done?

**Comment by hainm** [comment]
@arose  Can you please provide a code spinet for this so I could introduce to nglview? thanks.

**Comment by hainm** [comment]
hi @satary,

It turns out that `nglview` has not supported that feature yet. But fortunately, you can use the workaround (taken from: https://github.com/arose/ngl/blob/master/examples/scripts/color/volume.js)
[volume_color.ipynb.txt](https://github.com/arose/nglview/files/3025142/volume_color.ipynb.txt)

<img width="574" alt="Screen Shot 2019-03-30 at 12 33 41 AM" src="https://user-images.githubusercontent.com/4451957/55271492-ccb33200-5283-11e9-901a-84e8a49a1411.png">

**Comment by satary** [comment]
Hi @hainm!
Thanks for the suggestion, I will try to use it and report!

**Comment by satary** [comment]
Hi @hainm, am I missing something, or method _execute_js_code() is not in the current release?

**Comment by hainm** [comment]
I think it is in 1.1.9

Hai

**Comment by hainm** [comment]
Have you tried my attached notebook above yet? It works for me.

I will post another example soon, not at my computer right now.

Hai

**Comment by satary** [comment]
Yeap, I've tried it. No color change for me =(

**Comment by hainm** [comment]
Uhm

I just saw your editted comment (it would be better if you can post a new
post so I can get notification): can you summarize again what doesn’t work
and in which environment?

Just try with notebook first. Jupyterlab is not that stable (in my opinion).

To make sure the _execute_js_code work, just try a simple JavaScript code

code = “””
console.log(‘hello’)
“””

Call the method and open the JavaScript console to see the ‘hello’ word.

(Right click -> Inspection (or similar))

**Comment by satary** [comment]
Thanks for the reply!
In jupyter notebook (run from jupyter hub) I tested your code snippet and nothing seems to be in the console output. But! There is an info  message `widget_ngl.js:1010 can not create func for execute_code`
![image](https://user-images.githubusercontent.com/2101895/56309270-e2fd3100-6151-11e9-8bb5-e66dbb6fc3fd.png)

**Comment by satary** [comment]
As for the jupyterlab support there is a bug, I believe as it can not load any files with `add_component` method. 
![](https://user-images.githubusercontent.com/2101895/56287102-3193d680-6124-11e9-8340-2d32beadfee9.png)

**Comment by satary** [comment]
Update, I've managed to get Hello in the console, It was due to some mismatch in js extension and python extension. Gosh, it is a bit messy.
Now I can color the surface! Will attach my example later!
Thank you!

**Comment by hainm** [comment]
![image](https://user-images.githubusercontent.com/4451957/56310628-5daa5a00-611a-11e9-9dc9-d2487a023f52.png)

**Comment by satary** [comment]
To recap for the Google bots:
My task was to color molecular surface according to electostatic potential map (calculated by APBS) loaded from .dx file.
Here is a code snippet for that.
```
view=nv.show_structure_file('1KX5.pdb')
view.add_component('resources/pot.dx')
view

code = """
var ms = this.stage.compList[0];
var potential = this.stage.compList[1];

ms.addRepresentation('surface', {
    colorVolume: potential.volume,
    colorScheme: 'volume',
    colorScale: 'rwb',
    colorReverse: false,
    colorDomain: [ 0, 1500 ]
})

"""
view.component_1.clear_representations()
view.component_0.clear_representations()
view._execute_js_code(code)
 ```
![image](https://user-images.githubusercontent.com/2101895/56312226-66218580-6158-11e9-9d99-04193d5b87bb.png)

**Comment by hainm** [comment]
great. Thanks.

>

**Comment by hainm** [comment]
> I wonder if it is possible to color a structure (cartoon, surface and other atomic representations) by volume data (loaded with ccp4 or dx).

hi @satary, this feature is added to now #804

**Comment by satary** [comment]
Hello @hainm !
Great news and thank you a lot!
